nomad/csi_batch.go

```diff
@@ -0,0 +1,73 @@
+package nomad
+
+import (
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// csiBatchRelease is a helper for any time we need to release a bunch
+// of volume claims at once. It de-duplicates the volumes and batches
+// the raft messages into manageable chunks. Intended for use by RPCs
+// that have already been forwarded to the leader.
+type csiBatchRelease struct {
+	srv    *Server
+	logger log.Logger
+
+	maxBatchSize int
+	seen         map[string]struct{}
+	batches      []*structs.CSIVolumeClaimBatchRequest
+}
+
+func newCSIBatchRelease(srv *Server, logger log.Logger, max int) *csiBatchRelease {
+	return &csiBatchRelease{
+		srv:          srv,
+		logger:       logger,
+		maxBatchSize: max,
+		seen:         map[string]struct{}{},
+		batches:      []*structs.CSIVolumeClaimBatchRequest{{}},
+	}
+}
+
+// add the volume ID + namespace to the deduplicated batches
+func (c *csiBatchRelease) add(vol, namespace string) {
+	id := vol + namespace
+
+	// ignore duplicates
+	_, seen := c.seen[id]
+	if seen {
+		return
+	}
+
+	req := structs.CSIVolumeClaimRequest{
+		VolumeID: vol,
+		Claim:    structs.CSIVolumeClaimRelease,
+	}
+	req.Namespace = namespace
+	req.Region = c.srv.config.Region
+
+	for _, batch := range c.batches {
+		// otherwise append to the first non-full batch
+		if len(batch.Claims) < c.maxBatchSize {
+			batch.Claims = append(batch.Claims, req)
+			return
+		}
+	}
+	// no non-full batch found, make a new one
+	newBatch := &structs.CSIVolumeClaimBatchRequest{
+		Claims: []structs.CSIVolumeClaimRequest{req}}
+	c.batches = append(c.batches, newBatch)
+}
+
+// apply flushes the batches to raft
+func (c *csiBatchRelease) apply() error {
+	for _, batch := range c.batches {
+		if len(batch.Claims) > 0 {
+			_, _, err := c.srv.raftApply(structs.CSIVolumeClaimBatchRequestType, batch)
+			if err != nil {
+				c.logger.Error("csi raft apply failed", "error", err, "method", "claim")
+				return err
```

Comment:
first comment: the error returned from this method isn't consumed by either of the current calls (`Job.Deregister` and `Node.UpdateAlloc`).

second comment: what happens to volume claims in later batches if an intermediate batch has an apply error? is it worth holding the errors (perhaps in a multierror) and moving through the rest of the batches? are these reaped in a different GC procedure?

Reply:
I think the approach to pick things up in a multierror is right; the claims do eventually get GC’d but making a best effort to get them as many as we can in the first pass is a good optimization.